### ISSUE TITLE
Loss gets nan

### ISSUE THREAD

**Issue by yezhangyinge** [Xiaobin-Rong/gtcrn#11]
I use the loss function in loss.py to train my network, but I get nan in some epoch, how can i fix this problem?

**Comment by Xiaobin-Rong** [comment]
It shouldn't occur. I'd be glad to offer some advice if you can provide more details.

**Comment by TungyuYoung** [comment]
While I used the optimizer and the shceduler that the author suggest, the same problem occured. I think that there may be some of reasons:
1. The dataset may be not high quality enough
2. Gradient vanishing or exploding maight happen

To solve this problem, I use the **StepLR** instead of the original scheduler. And, a effective LR adjustment strategy should be considered.
@yezhangyinge @Xiaobin-Rong

**Comment by Xiaobin-Rong** [comment]
@TungyuYoung 
Thanks for sharing your experience. So, by adopting the **StepLR** scheduler, the **NAN** loss problem was solved? BTW, `torch.nn.utils.clip_grad_norm_` should be used to avoid gradient explosion.

**Comment by TungyuYoung** [comment]
> @TungyuYoung Thanks for sharing your experience. So, by adopting the **StepLR** scheduler, the **NAN** loss problem was solved? BTW, `torch.nn.utils.clip_grad_norm_` should be used to avoid gradient explosion.

That's correct. Note that I use a very large dataset not only DNS 3 but also AiShell's. And the noise dataset is also very large. I set the training set to 2000 hours and the test set to 200 hours. The NAN loss issue occurred around the 7th epoch. If use the scheduler you suggested, once NAN values occur, it is impossible to return to normal even if the learning rate is adjusted to a very small value. I did not use `torch.nn.utils.clip_grad_norm_ `. So I think that the root cause is the occurrence of gradient vanishing, which may be due to excessive use of GRU. What's more, the LR adjustment strategy I used is that it will be times a factor if the current test loss is smaller than the last epoch's (the factor can be set to any small decimal less than 1. I typically set it to 0.5).

**Comment by TungyuYoung** [comment]
> > @TungyuYoung Thanks for sharing your experience. So, by adopting the **StepLR** scheduler, the **NAN** loss problem was solved? BTW, `torch.nn.utils.clip_grad_norm_` should be used to avoid gradient explosion.
> 
> That's correct. Note that I use a very large dataset not only DNS 3 but also AiShell's. And the noise dataset is also very large. I set the training set to 2000 hours and the test set to 200 hours. The NAN loss issue occurred around the 7th epoch. If use the scheduler you suggested, once NAN values occur, it is impossible to return to normal even if the learning rate is adjusted to a very small value. I did not use `torch.nn.utils.clip_grad_norm_ `. So I think that the root cause is the occurrence of gradient vanishing, which may be due to excessive use of GRU. What's more, the LR adjustment strategy I used is that it will be times a factor if the current test loss is smaller than the last epoch's (the factor can be set to any small decimal less than 1. I typically set it to 0.5).

By the way, I have added a very small value to all divisions or square roots when calculating the loss to avoid issues.

**Comment by Xiaobin-Rong** [comment]
@TungyuYoung 
Thanks! I have also used the **stepLR** scheduler and found it better when the dataset is relatively large. Concretely, the learning rate is multiplied by a factor of 0.5 at epochs 80, 120, 150, 170, 180, 190, with a total training epoch of 200.
On the other hand, even though you have addressed the NAN loss issue, I would still recommend using gradient clipping during training.

**Comment by yezhangyinge** [comment]
> @TungyuYoung Thanks for sharing your experience. So, by adopting the **StepLR** scheduler, the **NAN** loss problem was solved? BTW, `torch.nn.utils.clip_grad_norm_` should be used to avoid gradient explosion.

actually, I fixed this problem by changing the si-snr loss. I found that you use this code:
`sisnr =  - torch.log10(torch.norm(y_true, dim=-1, keepdim=True)**2 / torch.norm(y_pred - y_true, dim=-1, keepdim=True)**2 + 1e-8).mean()`
When y_pred == y_true, loss will become nan maybe. : )

**Comment by yezhangyinge** [comment]
> While I used the optimizer and the shceduler that the author suggest, the same problem occured. I think that there may be some of reasons:
> 
> 1. The dataset may be not high quality enough
> 2. Gradient vanishing or exploding maight happen
> 
> To solve this problem, I use the **StepLR** instead of the original scheduler. And, a effective LR adjustment strategy should be considered. @yezhangyinge @Xiaobin-Rong

the true reason maybe this:
> 1. The dataset may be not high quality enough

**Comment by Xiaobin-Rong** [comment]
> > @TungyuYoung Thanks for sharing your experience. So, by adopting the **StepLR** scheduler, the **NAN** loss problem was solved? BTW, `torch.nn.utils.clip_grad_norm_` should be used to avoid gradient explosion.
> 
> actually, I fixed this problem by changing the si-snr loss. I found that you use this code: `sisnr = - torch.log10(torch.norm(y_true, dim=-1, keepdim=True)**2 / torch.norm(y_pred - y_true, dim=-1, keepdim=True)**2 + 1e-8).mean()` When y_pred == y_true, loss will become nan maybe. : )

Maybe you're right. However, I'm skeptical about whether the scenario where y_pred equals y_true actually occurs in reality.

**Comment by yezhangyinge** [comment]
> > > @TungyuYoung Thanks for sharing your experience. So, by adopting the **StepLR** scheduler, the **NAN** loss problem was solved? BTW, `torch.nn.utils.clip_grad_norm_` should be used to avoid gradient explosion.
> > 
> > 
> > actually, I fixed this problem by changing the si-snr loss. I found that you use this code: `sisnr = - torch.log10(torch.norm(y_true, dim=-1, keepdim=True)**2 / torch.norm(y_pred - y_true, dim=-1, keepdim=True)**2 + 1e-8).mean()` When y_pred == y_true, loss will become nan maybe. : )
> 
> Maybe you're right. However, I'm skeptical about whether the scenario where y_pred equals y_true actually occurs in reality.

Hmm, I am not sure if this will happen in reality but I just met. So I think maybe my dataset is not high quality enough, so this problem happened. Thank you very much for you reply!

**Comment by TungyuYoung** [comment]
> > > > @TungyuYoung Thanks for sharing your experience. So, by adopting the **StepLR** scheduler, the **NAN** loss problem was solved? BTW, `torch.nn.utils.clip_grad_norm_` should be used to avoid gradient explosion.
> > > 
> > > 
> > > actually, I fixed this problem by changing the si-snr loss. I found that you use this code: `sisnr = - torch.log10(torch.norm(y_true, dim=-1, keepdim=True)**2 / torch.norm(y_pred - y_true, dim=-1, keepdim=True)**2 + 1e-8).mean()` When y_pred == y_true, loss will become nan maybe. : )
> > 
> > 
> > Maybe you're right. However, I'm skeptical about whether the scenario where y_pred equals y_true actually occurs in reality.
> 
> Hmm, I am not sure if this will happen in reality but I just met. So I think maybe my dataset is not high quality enough, so this problem happened. Thank you very much for you reply!


When the NAN value issue first occurred, I attempted to replicate the problem and printed out each individual loss value. I observed that the si-snr loss was the first to exhibit a NAN value. I then saved the corresponding audio and found that it was completely blank. @Xiaobin-Rong  However, even after fixing the dataset and loss function, NAN values still appeared after training for a while. Therefore,  I adjusted the scheduler, and the NAN value issue no longer occurred.

So, my suggestion is: 
1. further clean the dataset or add a function to check for blank audio before inputting the data. 
2. Secondly, whether the scheduler needs adjustment is still worth considering.

**Comment by Xiaobin-Rong** [comment]
@TungyuYoung 
Thank you for your comprehensive reply! Your suggestion is really useful and insightful. The reason I didn't encounter this problem might be due to the long length of the training data. I set each utterance to a length of 8 seconds to prevent the audio from being completely silent.

**Comment by yezhangyinge** [comment]
> > > > > @TungyuYoung Thanks for sharing your experience. So, by adopting the **StepLR** scheduler, the **NAN** loss problem was solved? BTW, `torch.nn.utils.clip_grad_norm_` should be used to avoid gradient explosion.
> > > > 
> > > > 
> > > > actually, I fixed this problem by changing the si-snr loss. I found that you use this code: `sisnr = - torch.log10(torch.norm(y_true, dim=-1, keepdim=True)**2 / torch.norm(y_pred - y_true, dim=-1, keepdim=True)**2 + 1e-8).mean()` When y_pred == y_true, loss will become nan maybe. : )
> > > 
> > > 
> > > Maybe you're right. However, I'm skeptical about whether the scenario where y_pred equals y_true actually occurs in reality.
> > 
> > 
> > Hmm, I am not sure if this will happen in reality but I just met. So I think maybe my dataset is not high quality enough, so this problem happened. Thank you very much for you reply!
> 
> When the NAN value issue first occurred, I attempted to replicate the problem and printed out each individual loss value. I observed that the si-snr loss was the first to exhibit a NAN value. I then saved the corresponding audio and found that it was completely blank. @Xiaobin-Rong However, even after fixing the dataset and loss function, NAN values still appeared after training for a while. Therefore, I adjusted the scheduler, and the NAN value issue no longer occurred.
> 
> So, my suggestion is:
> 
> 1. further clean the dataset or add a function to check for blank audio before inputting the data.
> 2. Secondly, whether the scheduler needs adjustment is still worth considering.

Yes, you are right! I had also found that the audio is blank. But NAN values do not appeared after training for a while in my experiments. What I suggest is to fix the loss function like this. Maybe it will no longer get NAN. : )
`sisnr =  - torch.log10(torch.norm(y_true, dim=-1, keepdim=True)**2 / (torch.norm(y_pred - y_true, dim=-1, keepdim=True)**2+1e-8) + 1e-8).mean()`

**Comment by Xiaobin-Rong** [comment]
@yezhangyinge 
I have modified the code as you suggested, thank you!
